---
title: "PH525.5x Section 1: What we measure and why"
author: "Lauren Kemperman"
pdf_document: default
---
# Section 1: What we measure and why 

## Mammaprint Gene Signature
* Exploring genes used in the Mammaprint gene signature - assess risk of breast cancer 
* Diagnostic signature using gene expression levels of 70 genes
* Information about the 70 gene signature used in the Mammaprint algorithm 

```{r}
library(genefu)
data(sig.gene70)
dim(sig.gene70)
head(sig.gene70)[,1:6]
count_nan_gene_symbol <- sum(is.na(sig.gene70$NCBI.gene.symbol))
paste("Count of NaN NCBI gene symbols: ", count_nan_gene_symbol)
subset_matching_desc <- sig.gene70[which(sig.gene70$Description == "cyclin E2"), ]
paste("NCBI gene matching the description cyclin E2: ", subset_matching_desc$NCBI.gene.symbol)
number_kinase_coding_genes <- length(grep("kinase", sig.gene70$Description))
paste("Number of kinase coding genes responsible for cell to cell communication: ", number_kinase_coding_genes)
```

## Assessment: Phenotypes 
* COPDSexualDimorphism.data package - phenotypes (cols) individuals (rows)
* Data to assess incidence of COPD and emphysema by gender and smoking status
* The pkyrs variable in the expr.meta data.frame represents pack years smoked. Other variables include gender and diagmaj (disease status). These variables correspond to phenotypes.

```{r}
library(COPDSexualDimorphism.data)
data(lgrc.expr.meta)
head(expr.meta)
table(expr.meta$GENDER)
summary(expr.meta$pkyrs)

qqnorm(expr.meta$pkyrs, pch=1, frame=FALSE)
qqline(expr.meta$pkyrs, col = "steelblue", lwd = 2)

boxplot(pkyrs~gender, data=expr.meta)

```

## Assessment: Chromosomes and SNPs 
* GWAS (Genome-wide association studies)
* Comparing individuals with disease vs. controls using SNP chips or DNA sequencing. 
* SNPs with association are investigated for disruption of gene regulation or function 
* Bioconductor *gwascat* package

```{r}
library(gwascat)
data(ebicat_2020_04_30)
ebicat_2020_04_30

sort(table(ebicat_2020_04_30$CHR_ID), decreasing=TRUE)
```

## Microarray Technology 1: How Hybridization Works 
* Two technologies: microarray and NGS 
* Both counting DNA or RNA molecules 
* Both use a trick which allows us to take double-stranded DNA and convert to single-stranded 
* Both require thousands - millions of molecules for us to be able to measure anything 
* If a few cells only, they must be amplified 

### Microarray Technology 
1. Denaturation (single-stranded)
2. Hybridization - when you have a single strand in solution and it finds complimentary DNA, it will hybridize to form 2 stranded DNA. This can be exploited to count molecules 
3. Can create probes / troughs for different sequences. Put on location on piece of solid for the molecules we want to be able to count. Probes have compliments to the DNA that we want to count.

#### How microarray technology works 
* Piece of solid where we put probes - 1x1 cm piece of silicone that gets divided into thousands to millions of cells (difference squares)
* squares correspond to probes which represent molecules we are trying to count 
* 25bP long probes in example 
* second step: label a sample with fluorescent tags and put on array. hope that right molecules hybridize to right probes

#### Two-color microarrays 
* Hybridize two samples onto one array - two different labels that scanner can recognize 
* Advantages: cost savings 
* Sample 1: color 1, Sample 2: color 2. Let hybridize and get both hybridized to same probes, but scanner can distinguish two types of labels. 
* Two numbers per probe -- converted into RGB color combining red and green 

#### Applications of microarray technology 
* 3 different applications 

1. **Measuring gene expression** - gene chip array. 

* For every gene, we know the sequence and take 11 sequences for individual transcripts and hybridize. 
* On this array, probes are towards 3' end of transcripts b/c RNA tends to degrade more on one side (5' end). 
* 11 probes scattered around array to avoid confounding location with gene for each transcript. 
* Label the RNA, put it on the array. Will see lots of hybridization if there are many copies of that transcript. 
* High intensity = highly expressed gene. For each gene, select n probes and put them on the array and analyze the data. 

2. **Genotyping SNP** - different alleles 2 of same or 1 of each.

* I.e., AA, AG, GG. 
* If we want to know which of the three possibilities, we can do this for SNPs. 
* Use probes to hybridize to piece of sequence which has A, G for example. 
* Genotype millions of SNPs at a time. Arrays popular for GWAS studies to understand which alleles are associated with genes of interest. 

3. **Detection of transcription factor binding sites** - genome is more than just sequence, measuring the chemical processes taking place around the genome, i.e. where specific protein is bound. 

* Transcription factor = proteins that start gene expression. * Have DNA, want to know where specific protein is bound. Start by fragmenting DNA, some pieces have protein and others do not. 
* Divide by presence of protein vs. not - hybridize the part with protein with tiling array and if lights up, the location is where the protein was bound. 
* Intensities are not that reliable, must be controlled by hybridizing the total DNA for comparison. 



### Labeling 
Need indirect ways to count molecules. Labeling adds a chemical to each molecule, use optical scanner to identify the different intensities based on # labels and quantify. 

Design attribute of different technologies: synthetically sequenced, or cloned. Densities of probes put on the solid is also variable across different technologies. Also # samples on each array differs. 
Major manufacturers: 

1. Affymetrix (high density, one color)
2. Agilent (circles on grid, one or two color)
3. Illumina (high density, one or two color)
   * Uses beads instead of in-situ sequencing
   
   
### Brief introduction of NGS 

* Early 21st century Human Genome sequenced 
* Took DNA from several humans, pooled together and sequenced base x base the entire thing (1st generation)
* Back then, millions of clones 1k bP long. Different labs would sequence each clone, and then put together using computational methods. Cost billions of dollars 
* NGS is high throughput - billions of bP in a week for thousands of dollars. 
* Many copies of DNA to obtain measurement.

1. Fragment DNA using mol bio --> feed into NGS sequencer 
2. Read sequences for all of the fragments to get the reads (bases) 

Illumina flow cells - 8 lanes (1 per sample)

* For each lane, we get 160mill short reads (50-70bP long)
* Starts with DNA sample with fragments. Add adapters to each one to allow fragments to attach to pieces of solid. Once attached, amplify each one so that we have millions of copies (clusters) - adapter + fragment + copies. 
* Add labeled nucleotides. Different from microArray - not having two molecules join. 
* First base of the sequence attached to compliment, and starts forming double stranded. Once the first nucleotide attaches, take a picture and read intensity of labels. Keep doing this until we get through almost the entire fragment. 
* Images from sequencing machine represent clusters - first base of sequence represented by cluster. 
* Next step we get another image corresponding to bases on the second location and assemble the whole molecular sequence. 

### Applications of NGS 

* Similarly to microArray technology, we can measure gene expression, genotype, location of transcription factor binding sites 
* First application: sequencing the genome 
* Resequencing: do not sequence the whole genome with a new subject of same species, only areas of interest. SNP discovery, genotyping, variant discovery and quantification 
* Measuring methalation 
* Used be 1000s genomes project, human epigenome project

**Going from series of reads to measurements**
* First step in analyzing the NGS sequences: finding where the reads came from. All reads get mapped to the genome : matches to the reference genome 
* First application: Variant detection 
* Finding new SNPs. Take sample, sequence, align to genome, go to any specific location and ask if it is a SNP by analyzing whether there are G's and A's (heterozygous) 
* There are sequencing errors 
* Deletion: alleles are missing a base 

* RNA seq - NGS to quantify gene expression 
* We have RNA - two samples to see if they are different. 
* RNA --> DNA, sequence DNA and map / align to reference genome 
* Compare gene expression in the two samples 

* ChipSeq - finding transcription factor binding sites 
* DNA --> separate fragments bound to specific proteins and sequence those sections 
* Location of genome bound to many reads --> means that it was bound to that protein 
* Peak Detectors to identify locations of protein binding sites 

***Analyzing gene expression microarray dataset***
```{r}
library(tissuesGeneExpression)
data(tissuesGeneExpression)
paste("log scale intensities for microarray probes: ")
head(e[,1:5])
paste("tissue types of each sample: ")
table(tissue)

paste("overall mean expression of 209169_at", mean(e["209169_at",]))
paste("mean expression by tissue: ")
sort(by(e["209169_at",], tissue, mean))
```

***Gene Associated with probe ID***
```{r}
library(hgu133a.db)
symbol = mapIds(hgu133a.db, keys=rownames(e), column="SYMBOL", keytype="PROBEID")
paste("gene associated with probe ID 209169_at", symbol["209169_at"])

num_features <- sum(symbol == "H2AX", na.rm=TRUE)
paste("number of features measuring expression of H2AX: ", num_features)

paste("associated probes: ")
symbol[grep("H2AX", symbol)]

paste("comparing distributions across tissues: ")
boxplot(as.numeric(e["205436_s_at",])~tissue)

paste("finding gene specific to placenta: ")
IDs = c("201884_at", "209169_at", "206269_at", "207437_at", "219832_s_at", "212827_at")
sort(rowMeans(e[IDs, which(tissue == "placenta")]))
```


